modules/mbedtls/zephyr_init.c

```diff
@@ -122,6 +84,51 @@ mbedtls_ms_time_t mbedtls_ms_time(void)
 	return (mbedtls_ms_time_t)k_uptime_get();
 }
 
+#if defined(CONFIG_MBEDTLS_ZEPHYR_ENTROPY_POLL) || defined(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
+static int get_random_data(uint8_t *output, size_t output_size, bool cs_only)
+{
+	int ret = MBEDTLS_ERR_ENTROPY_SOURCE_FAILED;
+
+#if defined(CONFIG_CSPRNG_ENABLED)
+	ret = sys_csrand_get(output, output_size);
+	if (ret == 0) {
+		return 0;
+	}
+#endif /* CONFIG_CSPRNG_ENABLED */
+
+	if (!cs_only) {
+		sys_rand_get(output, output_size);
+		ret = 0;
+	}
+
+	return ret;
+}
+#endif /* CONFIG_MBEDTLS_ZEPHYR_ENTROPY_POLL || CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG */
+
+#if defined(CONFIG_MBEDTLS_ZEPHYR_ENTROPY_POLL)
+int mbedtls_hardware_poll(void *data, unsigned char *output, size_t len,
+			  size_t *olen)
+{
+	int ret;
+	uint16_t request_len = len > UINT16_MAX ? UINT16_MAX : len;
+
+	ARG_UNUSED(data);
+
+	if (output == NULL || olen == NULL || len == 0) {
+		return MBEDTLS_ERR_ENTROPY_SOURCE_FAILED;
+	}
+
+	ret = get_random_data(output, len, false);
```

Comment:
Noting that here (in `mbedtls_hardware_poll()`) we're completely dropping the use of `CONFIG_ENTROPY_HAS_DRIVER`/`entropy_get_entropy()` (replacing it by `CONFIG_CSPRNG_ENABLED`/`sys_csrand_get()`).
Is this intended and if so why?

Reply:
As for what it concerns with the kconfigs, `CSPRNG_ENABLED` is set whenever `ENTROPY_HAS_DRIVER` [is set](https://github.com/zephyrproject-rtos/zephyr/blob/95622e49b215bcd3ff61f666db6d74eedd4f5b51/subsys/random/Kconfig#L83) and there is no way to disable the `random` subsys so IMO they can be considered as equivalent.

I think that the behavior of `mbedtls_hardware_poll()` is not changed by the new implementation. In the old design the entropy device was polled first (if available) and then `sys_rand_get()` was used as backup. In the new fashion I try `sys_rand_get()` first (again, if available) and then I use `sys_rand_get()` as fallback.
In facts `sys_rand_get()` is implemented by:
- `zephyr/subsys/random/random_ctr_drbg.c` and
- `zephyr/subsys/random/random_entropy_device.c`.
Both implementations get the data from an entropy driver.

In conclusion I think that my implementation basically re-implements the same pattern that we had previously. Am I wrong?

Reply:
I don't think so. I was just checking with you the logic. All good!